shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -815,14 +815,33 @@ - (void)onOrientationPreferencesUpdated:(NSNotification*)notification {
     if (update == nil) {
       return;
     }
+    [self performOrientationUpdate:update.unsignedIntegerValue];
+  });
+}
 
-    NSUInteger new_preferences = update.unsignedIntegerValue;
-
-    if (new_preferences != _orientationPreferences) {
-      _orientationPreferences = new_preferences;
-      [UIViewController attemptRotationToDeviceOrientation];
+- (void)performOrientationUpdate:(UIInterfaceOrientationMask)new_preferences {
+  if (new_preferences != _orientationPreferences) {
+    _orientationPreferences = new_preferences;
+    [UIViewController attemptRotationToDeviceOrientation];
+
+    UIInterfaceOrientationMask currentInterfaceOrientation =
+        1 << [[UIApplication sharedApplication] statusBarOrientation];
+    if (!(_orientationPreferences & currentInterfaceOrientation)) {
+      // Force orientation switch if the current orientation is not allowed
+      if (_orientationPreferences & UIInterfaceOrientationMaskPortrait) {
+        [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortrait) forKey:@"orientation"];
```

Comment:
Can you also add a comment about the semi-legal nature of this call too please?  Potentially this might break in the future and we want to give someone a better chance of find it if it does.